Introduce 'wavm' to execute WebAssembly command line programs.

icefox opened the pull request:
$wavm echo.wast Hello World!
Hello World!

wavm takes a program as the first argument (and optionally -text or -binary
to specify a format) and executes the function main. If main takes arguments
wavm will pass in any extra arguments.  If main returns i32 wavm's will pass
along that as the exit code.

;; Example program that takes in arguments and exits with the value of argc
(module
  (export "main" $main)
  (func $main (param $argc i32) (param $argv i32) (result i32)
    (return (get_local $argc))
  )
)

Programs have access to the following I/O functions through the Emscripten intrinsics
* getc
* ungetc
* fread (stdin)
* fputc (stdout|stderr)
* fwrite (stdout|stderr)
* fflush (stdout|stderr)

Putting it all together wavm allows developers to build WebAssembly programs
for the command line.

Included are three example WebAssembly WAST programs.
* helloworld.wast => Outputs Hello World!
* echo.wast => echo's the command line arguments
* tee.wast => echo's stdin to stdout and stderr

With the capabilities provided a program could for example provide image
manipulation, reading in the image from stdin, manipulating it from command
line arguments and outputting to stdout all from within safety of the sandbox.

A WebAssembly image manipulation module could both be run as command line
application by exposing a main() and be loaded into a browser (that supports
WebAssembly) where a graphical frontend can be presented to users which
would use the modules API.

AndrewScheidecker commented:
I like it, though I have a few nits about formatting (I'll comment inline). The main think I'd like you to change is to replace Run with this program. You don't need to do everything Run does, like printing out the return value; just make it work on Test/Benchmark/a.out.wast and Test/zlib/zlib.wast. I think that should just require supporting "_main" as the exported function name, and a main function that doesn't take parameters.

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,114 @@
+	size_t lenstr = strlen(str);
+	size_t lensuffix = strlen(suffix);
+	if (lenstr < lensuffix)
+		return false;
```
Can you put parentheses around these if-bodies to be consistent with the rest of the code?

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,114 @@
+		auto main_argc = (uint32)argc - main_argc_start;
+		auto main_argv = (uint32)Runtime::vmGrowMemory((main_argc+1) * sizeof(uint32));
+		auto main_argvGlobal = &Runtime::instanceMemoryRef<uint32>(main_argv);
+		for (int i = main_argc_start; i < argc; ++i) {
```
Brace should be on the next line for consistency.

AndrewScheidecker commented on Test/wast/tee.wast:
```diff
@@ -0,0 +1,29 @@
+  (import $_get__stdin "env" "get__stdin" (param) (result i32))
+  (import $_get__stdout "env" "get__stdout" (param) (result i32))
+  (import $_get__stderr "env" "get__stderr" (param) (result i32))
+  (export "main" $main)
```
Is this export right? I would expect it to be an error, since the actual function definition exports "main", and does not bind $main. That might be a validation bug...

icefox commented:
Benchmark and zlib look like they might be broken at the moment? 

```
$./bin/Run -text Test/Benchmark/a.out.wast _main
_main threw exception: out of memory

$ ./bin/Run -text Test/zlib/zlib.wast _main
sizes: %d,%d
Killed
```

icefox commented:
- Added support for _main
- Cleaned up the argument parsing and added syntax checking with -c and incorporated outputting the AST also into wavm as just another argument.
- When the function isn't main I added support for passing arguments so you can do the really cool and handy following thing:
```
$ ./bin/wavm Test/spec/fac.wast --function fac-iter 5
fac-iter returned: I64(120)
```
- Updated the example section of the Readme

icefox commented:
Fixed the build, ready for the next round of review

AndrewScheidecker commented:
I think I'd still prefer PrintWAST to be a separate program rather than an "--outputAST" option on wavm. In the same spirit, I'd prefer Validate as a separate program rather than a "--check" option.

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,224 @@
+	const char* outputASTFilename = 0;
+
+	auto args = argv;
+	while (*++args) {
```
brace should be on newline

icefox replied:
Sorry about that, I must say this is the first time I have ever seen this brace style and while I have been trying to follow it, habits die hard and feel free to point any out.  Any easy way to add a style cop to travis?

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,224 @@
+			showHelp();
+			return EXIT_SUCCESS;
+		}
+		else {
```
This should be `else { break; }` to be consistent with formatting elsewhere

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,224 @@
+
+	Runtime::Value *parameters = nullptr;
+	auto function = module->functions[functionExport->second];
+	if(!strcmp(functionName, "main") || !strcmp(functionName, "_main"))
```
I think this would be more intuitive for users if it instead took this branch based on whether explicit `--function` argument was present or not.

icefox replied:
Nice call, much more readable to just check off functionName

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,224 @@
+			Runtime::Value mainParameters[2] = {main_argc, main_argv};
+			parameters = mainParameters;
+		}
+	} else {
```
for consistency should be
```
}
else
{
```

AndrewScheidecker commented on Source/Programs/wavm.cpp:
```diff
@@ -0,0 +1,224 @@
+				main_argvGlobal[i-main_argc_start] = vmAddress;
+			}
+			main_argvGlobal[main_argc] = 0;
+			Runtime::Value mainParameters[2] = {main_argc, main_argv};
```
This is a bug: it sets `parameters` to point to this array, and uses the pointer after the array goes out of scope. IMO just make parameters a `std::vector`.

icefox commented:
Moved PrintWAST back out and removed --check.  

The reasoning for --check is like bash -n, ruby -c, python -m, etc.  Built into the binary that runs programs have it exit before actually running anything.  Very handy for build scripts to verify files have no syntax errors and is something I have used for a long time.    Happy to add it as a pull request at a later time, don't want it holding up this and have removed it.
